File: Sources/FWObjC/Plugin/Picker/ImagePickerControllerImpl.m
//
//  ImagePickerControllerImpl.m
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

#import "ImagePickerControllerImpl.h"

#pragma mark - __FWImagePickerControllerImpl

@implementation __FWImagePickerControllerImpl

+ (__FWImagePickerControllerImpl *)sharedInstance
{
    static __FWImagePickerControllerImpl *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[__FWImagePickerControllerImpl alloc] init];
    });
    return instance;
}

- (void)viewController:(UIViewController *)viewController
         showImagePicker:(__FWImagePickerFilterType)filterType
          selectionLimit:(NSInteger)selectionLimit
           allowsEditing:(BOOL)allowsEditing
             customBlock:(void (^)(id _Nonnull))customBlock
              completion:(void (^)(NSArray * _Nonnull, NSArray * _Nonnull, BOOL))completion
{
    if (self.showsAlbumController) {
        __FWImageAlbumController *albumController = [self albumControllerWithFilterType:filterType];
        __weak __typeof__(self) self_weak_ = self;
        albumController.pickerControllerBlock = ^__FWImagePickerController * _Nonnull{
            __typeof__(self) self = self_weak_;
            return [self pickerControllerWithFilterType:filterType selectionLimit:selectionLimit allowsEditing:allowsEditing customBlock:customBlock completion:completion];
        };
        
        UINavigationController *navigationController = [[UINavigationController alloc] initWithRootViewController:albumController];
        navigationController.modalPresentationStyle = UIModalPresentationFullScreen;
        [viewController presentViewController:navigationController animated:YES completion:NULL];
    } else {
        __FWImagePickerController *pickerController = [self pickerControllerWithFilterType:filterType selectionLimit:selectionLimit allowsEditing:allowsEditing customBlock:customBlock completion:completion];
        __weak __typeof__(self) self_weak_ = self;
        pickerController.albumControllerBlock = ^__FWImageAlbumController * _Nonnull{
            __typeof__(self) self = self_weak_;
            return [self albumControllerWithFilterType:filterType];
        };
        [pickerController refreshWithFilterType:filterType];

        UINavigationController *navigationController = [[UINavigationController alloc] initWithRootViewController:pickerController];
        navigationController.modalPresentationStyle = UIModalPresentationFullScreen;
        [viewController presentViewController:navigationController animated:YES completion:nil];
    }
}

- (__FWImagePickerController *)pickerControllerWithFilterType:(__FWImagePickerFilterType)filterType
                                             selectionLimit:(NSInteger)selectionLimit
                                              allowsEditing:(BOOL)allowsEditing
                                                customBlock:(void (^)(id _Nonnull))customBlock
                                                 completion:(void (^)(NSArray * _Nonnull, NSArray * _Nonnull, BOOL))completion
{
    __FWImagePickerController *pickerController;
    if (self.pickerControllerBlock) {
        pickerController = self.pickerControllerBlock();
    } else {
        pickerController = [[__FWImagePickerController alloc] init];
    }
    pickerController.allowsMultipleSelection = selectionLimit != 1;
    pickerController.maximumSelectImageCount = selectionLimit > 0 ? selectionLimit : INT_MAX;
    pickerController.shouldRequestImage = YES;
    pickerController.filterType = filterType;
    __weak __typeof__(self) self_weak_ = self;
    pickerController.previewControllerBlock = ^__FWImagePickerPreviewController * _Nonnull{
        __typeof__(self) self = self_weak_;
        return [self previewControllerWithAllowsEditing:allowsEditing];
    };
    pickerController.didCancelPicking = ^{
        if (completion) completion(@[], @[], YES);
    };
    pickerController.didFinishPicking = ^(NSArray<__FWAsset *> * _Nonnull imagesAssetArray) {
        NSMutableArray *objects = [NSMutableArray array];
        NSMutableArray *results = [NSMutableArray array];
        [imagesAssetArray enumerateObjectsUsingBlock:^(__FWAsset *obj, NSUInteger idx, BOOL *stop) {
            if (obj.requestObject) {
                [objects addObject:obj.requestObject];
                [results addObject:obj.requestInfo ?: @{}];
            }
        }];
        if (completion) completion(objects.copy, results.copy, objects.count < 1);
    };
    
    if (self.customBlock) self.customBlock(pickerController);
    if (customBlock) customBlock(pickerController);
    return pickerController;
}

- (__FWImageAlbumController *)albumControllerWithFilterType:(__FWImagePickerFilterType)filterType
{
    __FWImageAlbumController *albumController;
    if (self.albumControllerBlock) {
        albumController = self.albumControllerBlock();
    } else {
        albumController = [[__FWImageAlbumController alloc] init];
        albumController.pickDefaultAlbumGroup = self.showsAlbumController;
    }
    albumController.contentType = [__FWImagePickerController albumContentTypeWithFilterType:filterType];
    return albumController;
}

- (__FWImagePickerPreviewController *)previewControllerWithAllowsEditing:(BOOL)allowsEditing
{
    __FWImagePickerPreviewController *previewController;
    if (self.previewControllerBlock) {
        previewController = self.previewControllerBlock();
    } else {
        previewController = [[__FWImagePickerPreviewController alloc] init];
    }
    previewController.showsEditButton = allowsEditing;
    if (!previewController.cropControllerBlock && self.cropControllerBlock) {
        previewController.cropControllerBlock = self.cropControllerBlock;
    }
    return previewController;
}

@end

File: Sources/FWUIKit/Plugin/Picker/ImagePickerControllerImpl.swift
//
//  ImagePickerControllerImpl.swift
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

import Photos
import PhotosUI
import UIKit
#if FWMacroSPM
@_spi(FW) import FWFramework
#endif

// MARK: - ImagePickerControllerImpl
/// 自定义图片选取插件
open class ImagePickerControllerImpl: NSObject, ImagePickerPlugin, @unchecked Sendable {
    // MARK: - Accessor
    /// 单例模式
    @objc(sharedInstance)
    public static let shared = ImagePickerControllerImpl()

    /// 是否显示相册列表控制器，默认为NO，点击titleView切换相册
    open var showsAlbumController: Bool = false

    /// 自定义相册列表控制器句柄，默认nil时使用自带控制器
    open var albumControllerBlock: (@MainActor @Sendable () -> ImageAlbumController)?

    /// 自定义图片预览控制器句柄，默认nil时使用自带控制器
    open var previewControllerBlock: (@MainActor @Sendable () -> ImagePickerPreviewController)?

    /// 自定义图片选取控制器句柄，默认nil时使用自带控制器
    open var pickerControllerBlock: (@MainActor @Sendable () -> ImagePickerController)?

    /// 自定义图片裁剪控制器句柄，预览控制器未自定义时生效，默认nil时使用自带控制器
    open var cropControllerBlock: (@MainActor @Sendable (UIImage) -> ImageCropController)?

    /// 自定义视频导出质量，默认nil时不处理
    open var videoExportPreset: String?

    /// 图片选取全局自定义句柄，show方法自动调用
    open var customBlock: (@MainActor @Sendable (ImagePickerController) -> Void)?

    // MARK: - ImagePickerPlugin
    open func showImagePicker(
        filterType: ImagePickerFilterType,
        selectionLimit: Int,
        allowsEditing: Bool,
        customBlock: (@MainActor @Sendable (Any) -> Void)?,
        completion: @escaping @MainActor @Sendable ([Any], [Any], Bool) -> Void,
        in viewController: UIViewController
    ) {
        if showsAlbumController {
            let albumController = albumController(filterType: filterType)
            albumController.pickerControllerBlock = {
                ImagePickerControllerImpl.shared.pickerController(filterType: filterType, selectionLimit: selectionLimit, allowsEditing: allowsEditing, customBlock: customBlock, completion: completion)
            }

            let navigationController = UINavigationController(rootViewController: albumController)
            navigationController.modalPresentationStyle = .fullScreen
            viewController.present(navigationController, animated: true)
        } else {
            let pickerController = pickerController(filterType: filterType, selectionLimit: selectionLimit, allowsEditing: allowsEditing, customBlock: customBlock, completion: completion)
            pickerController.albumControllerBlock = {
                ImagePickerControllerImpl.shared.albumController(filterType: filterType)
            }
            pickerController.refresh(filterType: filterType)

            let navigationController = UINavigationController(rootViewController: pickerController)
            navigationController.modalPresentationStyle = .fullScreen
            viewController.present(navigationController, animated: true)
        }
    }

    // MARK: - Private
    private func pickerController(
        filterType: ImagePickerFilterType,
        selectionLimit: Int,
        allowsEditing: Bool,
        customBlock: (@MainActor @Sendable (Any) -> Void)?,
        completion: @escaping @MainActor @Sendable ([Any], [Any], Bool) -> Void
    ) -> ImagePickerController {
        var pickerController: ImagePickerController
        if let pickerControllerBlock {
            pickerController = pickerControllerBlock()
        } else {
            pickerController = ImagePickerController()
        }
        pickerController.allowsMultipleSelection = selectionLimit != 1
        pickerController.maximumSelectImageCount = selectionLimit > 0 ? UInt(selectionLimit) : UInt.max
        pickerController.shouldRequestImage = true
        pickerController.filterType = .init(rawValue: filterType.rawValue)
        pickerController.previewControllerBlock = {
            ImagePickerControllerImpl.shared.previewController(allowsEditing: allowsEditing)
        }
        pickerController.didCancelPicking = {
            completion([], [], true)
        }
        pickerController.didFinishPicking = { imagesAssetArray in
            var objects: [Any] = []
            var results: [Any] = []
            for imagesAsset in imagesAssetArray {
                if let requestObject = imagesAsset.requestObject {
                    objects.append(requestObject)
                    results.append(imagesAsset.requestInfo ?? [:])
                }
            }
            completion(objects, results, objects.count < 1)
        }

        if let videoExportPreset {
            pickerController.videoExportPreset = videoExportPreset
        }

        self.customBlock?(pickerController)
        customBlock?(pickerController)
        return pickerController
    }

    private func albumController(filterType: ImagePickerFilterType) -> ImageAlbumController {
        var albumController: ImageAlbumController
        if let albumControllerBlock {
            albumController = albumControllerBlock()
        } else {
            albumController = ImageAlbumController()
            albumController.pickDefaultAlbumGroup = showsAlbumController
        }
        albumController.contentType = ImagePickerController.albumContentType(filterType: filterType)
        return albumController
    }

    private func previewController(allowsEditing: Bool) -> ImagePickerPreviewController {
        var previewController: ImagePickerPreviewController
        if let previewControllerBlock {
            previewController = previewControllerBlock()
        } else {
            previewController = ImagePickerPreviewController()
        }
        previewController.showsEditButton = allowsEditing
        if previewController.cropControllerBlock == nil, cropControllerBlock != nil {
            previewController.cropControllerBlock = cropControllerBlock
        }
        return previewController
    }
}
